libgadget/petapm.c

```diff
@@ -367,6 +375,163 @@ void petapm_force(PetaPM * pm, petapm_prepare_func prepare,
+
+        //NOTE: The PetaPM structs for reionisation use the G variable for filter radius in order to use
+        //the transfer functions correctly
+        pm_mass->G = R;
+        pm_star->G = R;
+        pm_sfr->G = R;
+
+        //TODO: maybe allocate and free these outside the loop
+        pfft_complex * mass_filtered = (pfft_complex *) mymalloc("mass_filtered", pm_mass->priv->fftsize * sizeof(double));
+        pfft_complex * star_filtered = (pfft_complex *) mymalloc("star_filtered", pm_star->priv->fftsize * sizeof(double));
+        pfft_complex * sfr_filtered = (pfft_complex *) mymalloc("sfr_filtered", pm_sfr->priv->fftsize * sizeof(double));
+
+        /* apply the filtering at this radius */
+        pm_apply_transfer_function(pm_mass, mass_unfiltered, mass_filtered, transfer);
+        pm_apply_transfer_function(pm_star, star_unfiltered, star_filtered, transfer);
+        pm_apply_transfer_function(pm_sfr, sfr_unfiltered, sfr_filtered, transfer);
+        walltime_measure("/PMreion/calc");
+
+        double * star_real = (double * ) mymalloc2("star_real", pm_star->priv->fftsize * sizeof(double));
+        double * sfr_real = (double * ) mymalloc2("sfr_real", pm_sfr->priv->fftsize * sizeof(double));
+        
+        /* back to real space */
+        pfft_execute_dft_c2r(pm_mass->priv->plan_back, mass_filtered, mass_real);
+        pfft_execute_dft_c2r(pm_star->priv->plan_back, star_filtered, star_real);
+        pfft_execute_dft_c2r(pm_sfr->priv->plan_back, sfr_filtered, sfr_real);
+        walltime_measure("/PMreion/c2r");
+        
+        myfree(sfr_filtered);
+        myfree(star_filtered);
+        myfree(mass_filtered);
+
+        /* the reion loop calculates the J21 and stores it,
+         * for now the mass_real grid will be reused to hold J21
+         * on the last filtering step*/
+        reion_loop(pm_mass,pm_star,pm_sfr,mass_real,star_real,sfr_real,last_step);
+
+        /* since we don't need to readout star and sfr grids...*/
+        /* on the last step, the mass grid is populated with J21 and read out*/
+        myfree(sfr_real);
+        myfree(star_real);
+        
+        R = R / R_delta;
+    }
+    //J21 grid is exchanged to pm_mass buffer and freed
+    layout_build_and_exchange_cells_to_local(pm_mass, &pm_mass->priv->layout, pm_mass->priv->meshbuf, mass_real);
+    walltime_measure("/PMreion/comm");
+    //J21 read out to particles
+    pm_iterate(pm_mass, readout, regions, Nregions);
+    walltime_measure("/PMreion/readout");
+}
+
+/* We need a slightly different flow for reionisation, so I 
+ * will define these here instead of messing with the force functions*/
+void petapm_reion(PetaPM * pm_mass, PetaPM * pm_star, PetaPM * pm_sfr,
+        petapm_prepare_func prepare,
+        PetaPMGlobalFunctions * global_functions, //petapm_transfer_func global_transfer,
+        PetaPMFunctions * functions,
+        PetaPMParticleStruct * pstruct,
+        petapm_reion_func reion_loop,
```

Comment:
This function compares the filtered and inverse transformed mass, stellar mass and sfr grids to determine what is ionised and J21 in each cell. I couldn't fit this into the transfer or readout functions without some serious modification, since it is applied after the c2r, can't operate within the iterate functions and requires all three grids.

Reply:
I see, that makes perfect sense but could you add a comment here saying that?